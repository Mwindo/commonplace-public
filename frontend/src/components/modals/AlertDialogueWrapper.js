import classes from "./ModalWrapper.module.css";

function AlertDialogueWrapper({ children }) {
  // TODO: use SASS, inherit from the classes rather than setting style here, and manage z-indices in some shared place
  return (
    <div className={classes.modal_wrapper} style={{ zIndex: 18 }}>
      <div className={classes.modal_background} style={{ zIndex: 17 }}></div>
      <div role="alertdialog" className={classes.modal_box} style={{ zIndex: 18 }}>
        <div className={classes.modal_inner}>{children}</div>
      </div>
    </div>
  );
}

export default AlertDialogueWrapper;
