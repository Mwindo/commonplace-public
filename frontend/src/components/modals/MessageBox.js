import classes from "./MessageBox.module.css";

function MessageBox({
  message,
  onConfirm,
  onCancel,
  onConfirmText = "Confirm",
  onCancelText = "Cancel",
}) {
  return (
    <>
      <div className={classes.message_box_message}>{message}</div>
      <div className={classes.message_box_button_wrapper}>
        {onConfirm && (
          <button
            autoFocus
            aria-label="Confirm"
            className={classes.message_box_confirm}
            onClick={onConfirm}
          >
            {onConfirmText}
          </button>
        )}
        {onCancel && (
          <button
            aria-label="Cancel"
            className={classes.message_box_cancel}
            onClick={onCancel}
          >
            {onCancelText}
          </button>
        )}
      </div>
    </>
  );
}

export default MessageBox;
