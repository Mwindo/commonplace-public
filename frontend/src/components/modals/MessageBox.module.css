.message_box_message {
  margin-bottom: 2em;
}

.message_box_button_wrapper {
  display: flex;
  justify-content: space-evenly;
}

.message_box_confirm {
  background-color: white;
  font-size: 1em;
  border-radius: 0.25em;
  border: 1px solid black;
  transition: 0.2s;
  cursor: pointer;
  height: 2em;
}

.message_box_cancel {
  background-color: white;
  font-size: 1em;
  color: black;
  border-radius: 0.25em;
  border: 1px solid black;
  transition: 0.2s;
  cursor: pointer;
  height: 2em;
}

.message_box_confirm:hover {
  background-color: seagreen;
  color: white;
  font-size: 1.3em;
}

.message_box_cancel:hover {
  background-color: darkred;
  color: white;
  font-size: 1.3em;
}
