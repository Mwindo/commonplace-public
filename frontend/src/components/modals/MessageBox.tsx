import classes from "./MessageBox.module.css";

interface MessageBoxProps {
  message: string;
  onConfirm?: (() => void) | null;
  onCancel?: (() => void) | null;
  onConfirmText?: string;
  onCancelText?: string;
}

function MessageBox({
  message,
  onConfirm = null,
  onCancel = null,
  onConfirmText = "Confirm",
  onCancelText = "Cancel",
}: MessageBoxProps) {
  return (
    <>
      <div className={classes.message_box_message}>{message}</div>
      <div className={classes.message_box_button_wrapper}>
        {onConfirm && (
          <button
            autoFocus
            aria-label="Confirm"
            className={classes.message_box_confirm}
            onClick={onConfirm}
          >
            {onConfirmText}
          </button>
        )}
        {onCancel && (
          <button
            aria-label="Cancel"
            className={classes.message_box_cancel}
            onClick={onCancel}
          >
            {onCancelText}
          </button>
        )}
      </div>
    </>
  );
}

export default MessageBox;
