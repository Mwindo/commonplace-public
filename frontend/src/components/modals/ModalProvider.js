import { createContext, useCallback, useState } from "react";
import ModalWrapper from "./ModalWrapper";
import AlertDialogueWrapper from "./AlertDialogueWrapper";
import MessageBox from "./MessageBox";

export const ModalContext = createContext();

// TODO: Make more flexible as needed.
// E.g., rather than storing component directly, store state and rerender accordingly.
const ModalProvider = ({ children }) => {
  const [modal, setModal] = useState(null);
  const [alertDialogue, setAlertDialogue] = useState(null);

  const showModal = useCallback(
    (modal, onCancel = null, showCloseButton = false) => {
      setModal(
        <ModalWrapper onCancel={onCancel} showCloseButton={showCloseButton}>
          {modal}
        </ModalWrapper>
      );
    },
    []
  );

  const showAlertDialogue = useCallback(
    (dialogueComponent) => {
      setAlertDialogue(
        <AlertDialogueWrapper>{dialogueComponent}</AlertDialogueWrapper>
      );
    },
    [setAlertDialogue]
  );

  const closeModal = useCallback(() => {
    setModal(null);
  }, [setModal]);

  const closeAlertDialogue = useCallback(() => {
    setAlertDialogue(null);
  }, [setAlertDialogue]);

  const closeAllModals = useCallback(() => {
    closeModal();
    closeAlertDialogue();
  }, [closeModal, closeAlertDialogue]);

  const showErrorMessage = useCallback(
    (message = "There was an error :(") => {
      showAlertDialogue(
        <MessageBox
          message={message}
          onConfirm={closeAlertDialogue}
          onConfirmText="Ok"
        />
      );
    },
    [showAlertDialogue, closeAlertDialogue]
  );

  return (
    <ModalContext.Provider
      value={{
        modal,
        showModal,
        closeModal,
        alertDialogue,
        showAlertDialogue,
        closeAlertDialogue,
        closeAllModals,
        showErrorMessage,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};

export default ModalProvider;
