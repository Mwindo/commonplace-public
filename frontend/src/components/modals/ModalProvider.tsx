import { ReactNode, createContext, useCallback, useState } from "react";
import ModalWrapper from "./ModalWrapper";
import AlertDialogueWrapper from "./AlertDialogueWrapper";
import MessageBox from "./MessageBox";

interface ModalContextProps {
  modal: any,
  showModal: any,
  closeModal: any,
  alertDialogue: any,
  showAlertDialogue: any,
  closeAlertDialogue: any,
  closeAllModals: any,
  showErrorMessage: any
}

export const ModalContext = createContext<ModalContextProps>({} as ModalContextProps);

// TODO: Make more flexible as needed.
// E.g., rather than storing component directly, store state and rerender accordingly.
const ModalProvider = ({ children }: {children: ReactNode}) => {
  const [modal, setModal] = useState<ReactNode | null>(null);
  const [alertDialogue, setAlertDialogue] = useState<ReactNode | null>(null);

  const showModal = useCallback(
    (modal: ReactNode, onCancel = undefined, showCloseButton = false) => {
      setModal(
        <ModalWrapper onCancel={onCancel} showCloseButton={showCloseButton}>
          {modal}
        </ModalWrapper>
      );
    },
    []
  );

  const showAlertDialogue = useCallback(
    (dialogueComponent: ReactNode) => {
      setAlertDialogue(
        <AlertDialogueWrapper>{dialogueComponent}</AlertDialogueWrapper>
      );
    },
    [setAlertDialogue]
  );

  const closeModal = useCallback(() => {
    setModal(null);
  }, [setModal]);

  const closeAlertDialogue = useCallback(() => {
    setAlertDialogue(null);
  }, [setAlertDialogue]);

  const closeAllModals = useCallback(() => {
    closeModal();
    closeAlertDialogue();
  }, [closeModal, closeAlertDialogue]);

  const showErrorMessage = useCallback(
    (message = "There was an error :(") => {
      showAlertDialogue(
        <MessageBox
          message={message}
          onConfirm={closeAlertDialogue}
          onConfirmText="Ok"
        />
      );
    },
    [showAlertDialogue, closeAlertDialogue]
  );

  return (
    <ModalContext.Provider
      value={{
        modal,
        showModal,
        closeModal,
        alertDialogue,
        showAlertDialogue,
        closeAlertDialogue,
        closeAllModals,
        showErrorMessage,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};

export default ModalProvider;
