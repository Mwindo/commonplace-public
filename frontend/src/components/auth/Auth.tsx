import { ReactNode, createContext, useContext, useState } from "react";
import { gql } from "graphql-request";
import { useQuery } from "@tanstack/react-query";
import { getGQLQueryClient } from "../requests/gqlQueryClient";
import { ModalContext } from "../modals/ModalProvider";
import MessageBox from "../modals/MessageBox";
import { useNavigate } from "react-router-dom";

interface LoginContextProps {
  isAuth: boolean;
  setAuth: (auth: boolean) => void;
  login: ({
    username,
    password,
  }: {
    username: string;
    password: string;
  }) => Promise<LoginResponse>;
  logout: () => Promise<LogoutResponse | undefined>;
  showSessionExpired: () => void;
}

interface LoginResponse {
  login: {
    status: string;
    error: string;
    id: number;
  };
}

interface LogoutResponse {
  logout: boolean;
}

interface IsLoggedInResponse {
  is_logged_in: boolean;
}

export const LoginContext = createContext({} as LoginContextProps);

const IS_AUTH_CHECK_INTERVAL = 60000; // TODO: make more dynamic as needed

const loginMutationQuery = gql`
  mutation loginUser($username: String!, $password: String!) {
    login(username: $username, password: $password) {
      status
      error
      id
    }
  }
`;

const logoutMutationQuery = gql`
  mutation logout {
    logout
  }
`;

const loggedInQuery = gql`
  query isLoggedIn {
    is_logged_in
  }
`;

const LoginProvider = ({ children }: { children: ReactNode }) => {
  const [isAuth, setAuth] = useState(document.cookie ? true : false);

  const { showAlertDialogue, closeAllModals } = useContext(ModalContext);

  const login = async ({
    username,
    password,
  }: {
    username: string;
    password: string;
  }) => {
    const response: LoginResponse = await getGQLQueryClient().request(
      loginMutationQuery,
      {
        username: username,
        password: password,
      }
    );
    // TODO: Better error handling
    if (response["login"]["status"] === "true") {
      setAuth(true);
    } else {
      throw Error();
    }
    return response;
  };

  const navigate = useNavigate();

  const logout = async () => {
    if (!isAuth) {
      return;
    }
    const response: LogoutResponse = await getGQLQueryClient().request(
      logoutMutationQuery
    );
    if (response["logout"] === true) {
      setAuth(false);
    }
    navigate("/alligator", { state: { fromApp: true } });
    // window.location.reload(); // There is probably a more efficient way, but this ensures everything is properly reset
    return response;
  };

  const showSessionExpired = () => {
    setAuth(false);
    showAlertDialogue(
      <MessageBox
        message="Session has expired. Please log back in."
        onConfirmText="OK"
        onConfirm={() => {
          closeAllModals();
          logout();
        }}
      />
    );
  };

  const checkLoggedIn = async () => {
    const response: IsLoggedInResponse = await getGQLQueryClient().request(
      loggedInQuery
    );
    if (isAuth && response["is_logged_in"] === false) {
      showSessionExpired();
    }
    return response;
  };

  useQuery({
    queryKey: ["loggedInCheck"],
    queryFn: checkLoggedIn,
    refetchInterval: IS_AUTH_CHECK_INTERVAL,
    enabled: isAuth,
  });

  return (
    <LoginContext.Provider
      value={{ isAuth, setAuth, login, logout, showSessionExpired }}
    >
      {children}
    </LoginContext.Provider>
  );
};

export default LoginProvider;
