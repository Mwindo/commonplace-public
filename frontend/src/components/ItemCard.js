import { Link, useLocation } from "react-router-dom";
import classes from "./ItemCard.module.css";
import { useContext, useEffect, useState } from "react";
import { LoginContext } from "./auth/Auth";
import { ModalContext } from "./modals/ModalProvider";
import MessageBox from "./modals/MessageBox";
import AddOrEditItem from "./AddOrEditItem";
import { truncateText } from "./utilities/stringOperations";
import { ADD_ITEM_ID } from "./utilities/itemCardUtilities";

const getToolTipText = (itemID, title, description) => {
  if (itemID === ADD_ITEM_ID) return "Add New Item";
  return title + ": " + description;
};

const TITLE_MAX_NUM_CHARS = 54;

function ItemCard({
  itemId,
  itemImage,
  itemTitle,
  itemDescription,
  reloadItems = () => {},
  onDelete,
}) {
  /*
    A component to show a preview of the full item.
  */

  const { isAuth } = useContext(LoginContext);

  // When in the editing state, we show a modal.
  const [isEditing, setIsEditing] = useState(false);

  // Keep track of whether the modal edit form is dirty.
  const [hasEdits, setHasEdits] = useState(false);

  const {
    showModal,
    closeModal,
    closeAlertDialogue,
    showAlertDialogue,
    closeAllModals,
  } = useContext(ModalContext);

  useEffect(() => {
    if (!isEditing) return; // There is nothing to do if we aren't editing.

    // We show the add/edit component as a modal.
    showModal(
      <AddOrEditItem
        itemId={itemId}
        setDirty={setHasEdits}
        // When we save, we will show an alert dialogue to confirm.
        // Upon user confirmation, the we will close the modal and reload.
        // This means the component will rerender, and isEditing will be set to the default (false),
        // so there is no need to manually reset it here.
        onSave={() =>
          showAlertDialogue(
            <MessageBox
              message="Saved!"
              onConfirmText="OK"
              onConfirm={() => {
                reloadItems();
                closeAllModals();
              }}
            />
          )
        }
      />,
      // On cancel, we need to manually set isEditing to false. Otherwise, this useEffect will become useless.
      // If there were no edits, we go ahead and close without showing an alert.
      // If there were edits, make sure the user is ok losing them.
      () => {
        setIsEditing(false);
        if (!hasEdits) {
          closeModal();
          return;
        }
        showAlertDialogue(
          <MessageBox
            message="Leave without saving?"
            onConfirm={closeAllModals}
            onCancel={closeAlertDialogue}
            onConfirmText="Yes"
            onCancelText="No"
          />
        );
      },
      true // Show close button here so user has a visual sign that they can exit
    );
  }, [
    isEditing,
    hasEdits,
    reloadItems,
    itemId,
    closeAlertDialogue,
    closeAllModals,
    closeModal,
    showAlertDialogue,
    showModal,
  ]);

  // This would be a good place for useCallback if performance issues arise.
  const handleAddOrEditClicked = (e) => {
    e.preventDefault();
    setIsEditing(true);
  };

  // This would be a good place for useCallback if performance issues arise.
  const handleRemoveItemClicked = (e) => {
    e.preventDefault();
    showModal(
      <MessageBox
        message={"Are you sure you want to delete this item?"}
        onConfirm={() => {
          closeModal();
          onDelete();
        }}
        onCancel={closeModal}
        onConfirmText="Yes"
        onCancelText="No!"
      />
    );
  };

  /* This is sort of hacky, but:
    1) In the case of existing data, we render the ItemCard as a link to the Item's route.
    2) If the user is logged in, however, we insert a "dummy" card to "Add New Item."
    In this second case, we want everything to look the same, but to "deactivate" the link
    and have it show the Add/Edit modal instead. We "deactivate" by pointing the link to the
    same location, including query parameters.
  */
  const location = useLocation();
  // This would be a good place for useMemo if performance issues arise, or if this
  // logic were abstracted into a separate hook.
  const linkDestination =
    itemId > 0 ? `/item/${itemId}` : location.pathname + location.search;
  const onLinkClick = itemId === ADD_ITEM_ID ? handleAddOrEditClicked : null;

  return (
    <Link
      title={getToolTipText(itemId, itemTitle, itemDescription)}
      to={linkDestination}
      onClick={onLinkClick}
      className={classes.item}
    >
      <div role="listitem" className={classes.inner_card}>
        <div
          className={classes.inner_card_top}
          style={{ backgroundImage: `url(${itemImage}` }}
        >
          {
            // Show the edit/remove options if the user is logged in
            isAuth && itemId !== ADD_ITEM_ID && (
              <div role="menubar" className={classes.edit_button_container}>
                <button
                  aria-label="Edit Item"
                  onClick={handleAddOrEditClicked}
                  className={classes.edit_button}
                >
                  Edit
                </button>
                <button
                  aria-label="Delete Item"
                  onClick={handleRemoveItemClicked}
                  className={classes.remove_button}
                >
                  Delete
                </button>
              </div>
            )
          }
        </div>
        <div className={classes.inner_card_bottom}>
          <p>{truncateText(itemTitle, TITLE_MAX_NUM_CHARS)}</p>
        </div>
      </div>
    </Link>
  );
}

export default ItemCard;
