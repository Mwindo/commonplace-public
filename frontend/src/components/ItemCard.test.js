import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ItemCard from "./ItemCard";
import { LoginContext } from "./auth/Auth";
import { ModalContext } from "./modals/ModalProvider";
import { BrowserRouter } from "react-router-dom";

const testItemCard = ({
  isAuth,
  showModal = jest.fn(),
  showAlertDialogue = jest.fn(),
}) => {
  return (
    <ModalContext.Provider value={{ showModal, showAlertDialogue }}>
      <LoginContext.Provider value={{ isAuth }}>
        <BrowserRouter>
          <ItemCard
            itemId={1}
            itemImage=""
            itemTitle="Test Card"
            itemDescription="Test Description"
          />
        </BrowserRouter>
      </LoginContext.Provider>
    </ModalContext.Provider>
  );
};

it("check no edit buttons appear if not isAuth", () => {
  const { asFragment, queryByRole } = render(testItemCard({ isAuth: false }));
  expect(asFragment()).toMatchSnapshot();
  expect(queryByRole("menubar")).toBeFalsy();
});

it("check edit buttons appear if isAuth", () => {
  const { asFragment, queryByRole } = render(testItemCard({ isAuth: true }));
  expect(asFragment()).toMatchSnapshot();
  expect(queryByRole("menubar")).toBeTruthy();
});

it("check edit modal appears when edit clicked", async () => {
  let modalShown = false;
  const showModal = () => {
    modalShown = true;
  };
  const { getByText } = render(testItemCard({ isAuth: true, showModal }));
  await userEvent.click(getByText("Edit"));
  expect(modalShown).toBeTruthy();
});

it("check delete modal appears when delete clicked", async () => {
  let modalShown = false;
  const showModal = () => {
    modalShown = true;
  };
  const { getByText } = render(testItemCard({ isAuth: true, showModal }));
  await userEvent.click(getByText("Delete"));
  expect(modalShown).toBeTruthy();
});
