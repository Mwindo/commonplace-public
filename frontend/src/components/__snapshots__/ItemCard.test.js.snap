// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check edit buttons appear if isAuth 1`] = `
<DocumentFragment>
  <a
    class="item"
    href="/item/1"
    title="Test Card: Test Description"
  >
    <div
      class="inner_card"
      role="listitem"
    >
      <div
        class="inner_card_top"
      >
        <div
          class="edit_button_container"
          role="menubar"
        >
          <button
            aria-label="Edit Item"
            class="edit_button"
          >
            Edit
          </button>
          <button
            aria-label="Delete Item"
            class="remove_button"
          >
            Delete
          </button>
        </div>
      </div>
      <div
        class="inner_card_bottom"
      >
        <p>
          Test Card
        </p>
      </div>
    </div>
  </a>
</DocumentFragment>
`;

exports[`check no edit buttons appear if not isAuth 1`] = `
<DocumentFragment>
  <a
    class="item"
    href="/item/1"
    title="Test Card: Test Description"
  >
    <div
      class="inner_card"
      role="listitem"
    >
      <div
        class="inner_card_top"
      />
      <div
        class="inner_card_bottom"
      >
        <p>
          Test Card
        </p>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
